Improve public API

RobinMalfait opened the pull request:
This PR introduces a `public` folder inside the `src` folder. The idea here is that the files inside the `./src/public` folder will eventually be "public" API. The files inside the `./src/public` folder are written in `ESM`. This means that internally if we need those public files, we can then include them from there.
The files in the root will include them from the compiled `./lib/public` folder.

The advantage here is:
- We don't need to compile/transpile the code before running tests anymore, we can just run them.
  - Little caveat: We do still need to generate the plugin list, but that's located in a `pretest` script now, so you don't need to think about that!
- There is a clear set of files that will be exposed for public use. The vscode extension for example uses a few of those files already. This means that we technically only have to worry about backwards compatibility for these files without looking at the entire internet to see who is using private API's.

RobinMalfait committed:
introduce `public` folder

This can contain all of the `public` functions we want to expose.
This will be a bit nicer for example when you want to use
internal/private functions (we use some in the vscode intellisense
plugin).

RobinMalfait committed:
use public `resolveConfig` function

RobinMalfait committed:
expose resolveConfig in the root

This will use the resolveConfig we expose from the `public` folder. We
can probably generate these as well.

RobinMalfait committed:
make `colors` public

RobinMalfait committed:
make `default config` public

RobinMalfait committed:
make `default theme` public

RobinMalfait committed:
make `create plugin` public

adamwathan commented on resolveConfig.js:
```diff
@@ -1,7 +1 @@
-  const [, ...defaultConfigs] = getAllConfigs(configs[0])
-  return resolveConfigObjects([...configs, ...defaultConfigs])
-}
+module.exports = require('./lib/public/resolve-config').default
```
What's the reason for using dash-case here vs. camelCase for the other files?

RobinMalfait committed:
update to public paths

RobinMalfait committed:
remove `@tailwindcss/aspect-ratio` from tests

This should be tested in its own repo instead.

RobinMalfait committed:
remove `@tailwindcss/aspect-ratio` as a dependency

RobinMalfait committed:
drop `Build` step from CI

The build step is not a prerequisite anymore for running the tests. When
we want to release a new (insiders) release, the `prepublishOnly` step
will be executed for us.

Before this change, it would have been executed twice:
- Once before the tests
- Once before the actual release

RobinMalfait committed:
improve paths for caching purposes

RobinMalfait committed:
add pretest scrip for generating the plugin list

Now that we can use `SWC`, automatically generating the plugin list
before running the tests is super fast and you don't even have to think
about it anymore!
